Add Zookeeper Watcher

craigbr opened the pull request:
Also, add average round trip time, running average for last 5
round trips, and total number of round trips

craigbr committed:
Add Zookeeper Watcher

Also, add average round trip time, running average for last 5
round trips, and total number of round trips

mhoppal commented on main/zookeeper_watcher.go:
```diff
@@ -0,0 +1,207 @@
+//
```
// Copyright 2017 Hewlett Packard Enterprise Development LP

craigbr replied:
Done

mhoppal commented on main/zookeeper_watcher.go:
```diff
@@ -0,0 +1,207 @@
+	if len(zookeeperServers) < 0 {
+		log.Fatalf("Invalid ZOOKEEPER_SERVERS, must not be empty")
+	}
+	prometheusEndpoint := configuration.PrometheusEndpoint
```
why do you check all other variables are set correctly besides this one?

craigbr replied:
Done

mhoppal commented on main/zookeeper_watcher.go:
```diff
@@ -0,0 +1,207 @@
+	var eventChannel <-chan zk.Event
+	connection, eventChannel, err = connect(zookeeperServers)
+	if err != nil {
+		log.Fatalf("Connecting to Zookeeper failed: %s", err)
```
why exit why not emit a connect_status_metric?

mhoppal commented on main/zookeeper_watcher.go:
```diff
@@ -0,0 +1,207 @@
+		ev := <-eventChannel
+		log.Infof("Event = %s", ev)
+		if ev.State == zk.StateConnected {
+			log.Infof("Zookeeper state is connected")
```
and a connected state metric here!

craigbr committed:
Addressed review comments

Made it clear that connect just starts the connection process.

Add copyrights

matrixik commented on README.md:
```diff
@@ -32,22 +40,50 @@ Several parameters can be specified using environment variables:
 ## Metrics
 
-### Kafka Watcher
+### Zookeeper Watcher
```
Kafka

craigbr replied:
Done

matrixik commented on README.md:
```diff
@@ -1,17 +1,25 @@
+### Kafka Watcher
+
+The Kafka Watcher writes a message to a Kafka topic and then reads it back from the same
+topic. This ensures the entire round trip functionality of Zookeeper is functioning.
```
of Kafka?

craigbr replied:
Done

matrixik commented on README.md:
```diff
@@ -32,22 +40,50 @@ Several parameters can be specified using environment variables:
 ## Install (using Docker)
 
-See github.com/monasca/monasca-docker/kafka-watcher
+See github.com/monasca/monasca-docker/kafka-watcher and github.com/monasca/monasca-docker/zookeeper-watcher
```
Both of this links will show 404 error on github

craigbr replied:
Done

matrixik commented on main/zookeeper_watcher.go:
```diff
@@ -0,0 +1,212 @@
+	}
+
+	healthCheckPath := configuration.HealthCheckPath
+	if len(healthCheckPath) < 0 {
```
`< 1` or `<= 0`

matrixik commented on main/zookeeper_watcher.go:
```diff
@@ -0,0 +1,212 @@
+		healthCheckPath = "/" + healthCheckPath
+	}
+	zookeeperServers := configuration.ZookeeperServers
+	if len(zookeeperServers) < 0 {
```
`< 1` or `<= 0`

craigbr replied:
Should be == 0. Definitely not < 0. Done

matrixik commented on watcher/watcher.go:
```diff
@@ -104,29 +113,48 @@ func CreateWatcher(watchable Watchable, period time.Duration, maxFailures int, t
 	prometheus.MustRegister(droppedMessagesMetric)
 	prometheus.MustRegister(minRoundTripTimeMetric)
 	prometheus.MustRegister(maxRoundTripTimeMetric)
+	prometheus.MustRegister(averageRoundTripTimeMetric)
```
Register this in the same order like you initialized them in lines 116-128.

craigbr replied:
Good idea. Done

craigbr committed:
Resolved PR comments

Fix typos in README. Use good links.

Validate configuraion strings properly. Validate
PROMETHEUS_ENDPOINT.

Keep metric creation and registration in the same order to make
it easier to see issues.

craigbr commented:
@matrixik I've addressed all of your comments. Thanks for the review. Can you remove the "Changes requested"? Thanks
